src/hotspot/share/opto/runtime.cpp

```diff
@@ -132,13 +125,62 @@ static bool check_compiled_frame(JavaThread* thread) {
   var = generate_stub(env, type_func_gen, CAST_FROM_FN_PTR(address, c_func), #var, fancy_jump, pass_tls, return_pc); \
   if (var == nullptr) { return false; }
+*/
+
+#define GEN_OPTO_BLOB(name, type)                    \
+  generate_ ## name ## _blob();
+
+// a few helper macros to conjure up generate_stub call arguments
+#define OPTO_STUB_FIELD_NAME(name) _ ## name ## _Java
+#define OPTO_STUB_TYPEFUNC(name) name ## _Type
+#define OPTO_STUB_C_FUNC(name) CAST_FROM_FN_PTR(address, name ## _C)
+#define OPTO_STUB_NAME(name) stub_name(OptoStubId::name ## _id)
+
+// Almost all the C functions targeted from the generated stubs are
+// implemented locally to OptoRuntime with names that can be generated
+// from the stub name by appending suffix '_C'. However, in two cases
+// a common target method also needs to be called from shared runtime
+// stubs. In these two cases the opto stubs rely on method
+// imlementations defined in class SharedRuntime. The following
+// defines temporarily rebind the generated names to reference the
+// relevant implementations.
+
+#define complete_monitor_locking_C SharedRuntime::complete_monitor_locking_C
+#define slow_arraycopy_C SharedRuntime::slow_arraycopy_C
+
+#define GEN_OPTO_STUB(name, fancy_jump, pass_tls, pass_retpc  )         \
+  OPTO_STUB_FIELD_NAME(name) =                                          \
+    generate_stub(env,                                                  \
+                  OPTO_STUB_TYPEFUNC(name),                             \
+                  OPTO_STUB_C_FUNC(name),                               \
+                  OPTO_STUB_NAME(name),                                 \
+                  fancy_jump,                                           \
+                  pass_tls,                                             \
+                  pass_retpc);                                          \
+  if (OPTO_STUB_FIELD_NAME(name) == nullptr) { return false; }          \
+
+#define OPTO_JVMTI_STUB_C_FUNC(name) CAST_FROM_FN_PTR(address, SharedRuntime::name)
+
+#define GEN_OPTO_JVMTI_STUB(name)                                       \
+  STUB_FIELD_NAME(name) =                                               \
+    generate_stub(env,                                                  \
+                  notify_jvmti_vthread_Type,                            \
+                  OPTO_JVMTI_STUB_C_FUNC(name),                         \
+                  OPTO_STUB_NAME(name),                                 \
+                  0,                                                    \
+                  true,                                                 \
+                  false);                                               \
+  if (STUB_FIELD_NAME(name) == nullptr) { return false; }               \
 
 bool OptoRuntime::generate(ciEnv* env) {
 
+  OPTO_STUBS_DO(GEN_OPTO_BLOB, GEN_OPTO_STUB, GEN_OPTO_JVMTI_STUB)
+
+    /*
```

Comment:
This old code that has been commented out needs to be removed.

Reply:
Then do it.

Reply:
Done